File: odk/examples/DevelopersGuide/Charts/ChartInDraw.java
/* -*- Mode: Java; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- */
/*************************************************************************
 *
 *  The Contents of this file are made available subject to the terms of
 *  the BSD license.
 *
 *  Copyright 2000, 2010 Oracle and/or its affiliates.
 *  All rights reserved.
 *
 *  Redistribution and use in source and binary forms, with or without
 *  modification, are permitted provided that the following conditions
 *  are met:
 *  1. Redistributions of source code must retain the above copyright
 *     notice, this list of conditions and the following disclaimer.
 *  2. Redistributions in binary form must reproduce the above copyright
 *     notice, this list of conditions and the following disclaimer in the
 *     documentation and/or other materials provided with the distribution.
 *  3. Neither the name of Sun Microsystems, Inc. nor the names of its
 *     contributors may be used to endorse or promote products derived
 *     from this software without specific prior written permission.
 *
 *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
 *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
 *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
 *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
 *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
 *  OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 *  ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR
 *  TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
 *  USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 *************************************************************************/

// __________ Imports __________

import com.sun.star.uno.UnoRuntime;
import com.sun.star.lang.*;

// property access
import com.sun.star.beans.*;

// application specific classes
import com.sun.star.chart.*;
import com.sun.star.drawing.*;

import com.sun.star.frame.XModel;
// base graphics things
import com.sun.star.awt.Point;
import com.sun.star.awt.Size;
// Exceptions
import com.sun.star.uno.Exception;
import com.sun.star.uno.RuntimeException;
import com.sun.star.beans.UnknownPropertyException;


// __________ Implementation __________

// Create a spreadsheet add some data and add a chart

public class ChartInDraw
{


    public static void main( String args[] )
    {
        Helper aHelper = new Helper( args );

        ChartHelper aChartHelper = new ChartHelper( aHelper.createDrawingDocument());

        // the unit for measures is 1/100th of a millimeter
        // position at (1cm, 1cm)
        Point aPos    = new Point( 1000, 1000 );

        // size of the chart is 15cm x 12cm
        Size  aExtent = new Size( 15000, 13000 );

        // insert a new chart into the "Chart" sheet of the
        // spreadsheet document
        XChartDocument aChartDoc = aChartHelper.insertOLEChartInDraw(
            aPos,
            aExtent,
            "com.sun.star.chart.BarDiagram" );

        // instantiate test class with newly created chart
        ChartInDraw aTest   = new ChartInDraw( aChartDoc );

        try
        {
            aTest.lockControllers();

            aTest.testArea();
            aTest.testWall();
            aTest.testTitle();
            aTest.testLegend();
            aTest.testThreeD();

            aTest.unlockControllers();
        }
        catch( Exception ex )
        {
            System.out.println( "UNO Exception caught: " + ex );
            System.out.println( "Message: " + ex.getMessage() );
        }

        System.exit( 0 );
    }




    public ChartInDraw( XChartDocument aChartDoc )
    {
        maChartDocument = aChartDoc;
        maDiagram       = maChartDocument.getDiagram();
    }



    public void lockControllers()
        throws RuntimeException
    {
        UnoRuntime.queryInterface( XModel.class, maChartDocument ).lockControllers();
    }



    public void unlockControllers()
        throws RuntimeException
    {
        UnoRuntime.queryInterface( XModel.class, maChartDocument ).unlockControllers();
    }



    public void testArea()
        throws RuntimeException, UnknownPropertyException, PropertyVetoException,
               com.sun.star.lang.IllegalArgumentException, WrappedTargetException
    {
        XPropertySet   aArea = maChartDocument.getArea();

        if( aArea != null )
        {
            // change background color of entire chart
            aArea.setPropertyValue( "FillStyle", FillStyle.SOLID );
            aArea.setPropertyValue( "FillColor", Integer.valueOf( 0xeeeeee ));
        }
    }



    public void testWall()
        throws RuntimeException, UnknownPropertyException, PropertyVetoException,
               com.sun.star.lang.IllegalArgumentException, WrappedTargetException
    {
        XPropertySet aWall = UnoRuntime.queryInterface(
                                  X3DDisplay.class, maDiagram ).getWall();

        // change background color of area
        aWall.setPropertyValue( "FillColor", Integer.valueOf( 0xcccccc ));
        aWall.setPropertyValue( "FillStyle",  FillStyle.SOLID );
    }



    public void testTitle()
        throws RuntimeException, UnknownPropertyException, PropertyVetoException,
               com.sun.star.lang.IllegalArgumentException, WrappedTargetException
    {
        // change main title
        XPropertySet aDocProp = UnoRuntime.queryInterface(
            XPropertySet.class, maChartDocument );
        aDocProp.setPropertyValue( "HasMainTitle", Boolean.TRUE);

        XShape aTitle = maChartDocument.getTitle();
        XPropertySet aTitleProp = UnoRuntime.queryInterface( XPropertySet.class, aTitle );

        // set new text
        if( aTitleProp != null )
        {
            aTitleProp.setPropertyValue( "String", "Bar Chart in a Draw Document" );
        }
    }



    public void testLegend()
        throws RuntimeException, UnknownPropertyException, PropertyVetoException,
               com.sun.star.lang.IllegalArgumentException, WrappedTargetException
    {
        XShape aLegend = maChartDocument.getLegend();
        XPropertySet aLegendProp = UnoRuntime.queryInterface( XPropertySet.class, aLegend );

        aLegendProp.setPropertyValue( "Alignment", ChartLegendPosition.LEFT );
        aLegendProp.setPropertyValue( "FillStyle", FillStyle.SOLID );
        aLegendProp.setPropertyValue( "FillColor", Integer.valueOf( 0xeeddee ));
    }



    public void testThreeD()
        throws RuntimeException, UnknownPropertyException, PropertyVetoException,
        com.sun.star.lang.IllegalArgumentException, WrappedTargetException
    {
        XPropertySet aDiaProp = UnoRuntime.queryInterface( XPropertySet.class, maDiagram );
        Boolean aTrue = Boolean.TRUE;

        aDiaProp.setPropertyValue( "Dim3D", aTrue );
        aDiaProp.setPropertyValue( "Deep", aTrue );
        // from Chart3DBarProperties:
        aDiaProp.setPropertyValue( "SolidType", Integer.valueOf( ChartSolidType.CYLINDER ));

        // change floor color to Magenta6
        XPropertySet aFloor = UnoRuntime.queryInterface(
                                   X3DDisplay.class, maDiagram ).getFloor();
        aFloor.setPropertyValue( "FillColor", Integer.valueOf( 0x6b2394 ));

        // apply changes to get a 3d scene
        unlockControllers();
        lockControllers();


        // rotate scene to a different angle
        HomogenMatrix aMatrix = new HomogenMatrix();
        HomogenMatrixLine aLines[] = new HomogenMatrixLine[]
            {
                new HomogenMatrixLine( 1.0, 0.0, 0.0, 0.0 ),
                new HomogenMatrixLine( 0.0, 1.0, 0.0, 0.0 ),
                new HomogenMatrixLine( 0.0, 0.0, 1.0, 0.0 ),
                new HomogenMatrixLine( 0.0, 0.0, 0.0, 1.0 )
            };

        aMatrix.Line1 = aLines[ 0 ];
        aMatrix.Line2 = aLines[ 1 ];
        aMatrix.Line3 = aLines[ 2 ];
        aMatrix.Line4 = aLines[ 3 ];

        // rotate 10 degrees along the x axis
        double fAngle = 10.0;
        double fCosX = Math.cos( Math.PI / 180.0 * fAngle );
        double fSinX = Math.sin( Math.PI / 180.0 * fAngle );

        // rotate -20 degrees along the y axis
        fAngle = -20.0;
        double fCosY = Math.cos( Math.PI / 180.0 * fAngle );
        double fSinY = Math.sin( Math.PI / 180.0 * fAngle );

        // rotate -5 degrees along the z axis
        fAngle = -5.0;
        double fCosZ = Math.cos( Math.PI / 180.0 * fAngle );
        double fSinZ = Math.sin( Math.PI / 180.0 * fAngle );

        aMatrix.Line1.Column1 =  fCosY *  fCosZ;
        aMatrix.Line1.Column2 =  fCosY * -fSinZ;
        aMatrix.Line1.Column3 =  fSinY;

        aMatrix.Line2.Column1 =  fSinX *  fSinY *  fCosZ +  fCosX *  fSinZ;
        aMatrix.Line2.Column2 = -fSinX *  fSinY *  fSinZ +  fCosX *  fCosZ;
        aMatrix.Line2.Column3 = -fSinX *  fCosY;

        aMatrix.Line3.Column1 = -fCosX *  fSinY *  fCosZ +  fSinX *  fSinZ;
        aMatrix.Line3.Column2 =  fCosX *  fSinY *  fSinZ +  fSinX *  fCosZ;
        aMatrix.Line3.Column3 =  fCosX *  fCosY;

        aDiaProp.setPropertyValue( "D3DTransformMatrix", aMatrix );

        // add a red light source

        // in a chart by default only the second (non-specular) light source is switched on
        // light source 1 is a specular light source
        aDiaProp.setPropertyValue( "D3DSceneLightColor1", Integer.valueOf( 0xff3333 ));

        // set direction
        com.sun.star.drawing.Direction3D aDirection = new com.sun.star.drawing.Direction3D();

        aDirection.DirectionX = -0.75;
        aDirection.DirectionY =  0.5;
        aDirection.DirectionZ =  0.5;

        aDiaProp.setPropertyValue( "D3DSceneLightDirection1", aDirection );
        aDiaProp.setPropertyValue( "D3DSceneLightOn1", Boolean.TRUE);
    }



    // private members


    private final XChartDocument maChartDocument;
    private final XDiagram       maDiagram;
}

/* vim:set shiftwidth=4 softtabstop=4 expandtab: */

File: odk/examples/DevelopersGuide/Charts/python/ChartInDraw.py
# -*- tab-width: 4; indent-tabs-mode: nil; py-indent-offset: 4 -*-
#
# This file is part of the LibreOffice project.
#
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.
#

import math

import Helper
import ChartHelper

from com.sun.star.awt import Point, Size
from com.sun.star.drawing.FillStyle import SOLID
from com.sun.star.drawing import Direction3D, HomogenMatrix, HomogenMatrixLine
from com.sun.star.chart.ChartLegendPosition import LEFT
from com.sun.star.chart.ChartSolidType import CYLINDER


class ChartInDraw(object):

    def __init__(self, chart_doc):
        self._chart_document = chart_doc
        self._diagram = chart_doc.getDiagram()

    def lock_controllers(self):
        self._chart_document.lockControllers()

    def unlock_controllers(self):
        self._chart_document.unlockControllers()

    def test_area(self):
        area = self._chart_document.getArea()
        if area is not None:
            # change background color of entire chart
            area.setPropertyValue("FillStyle", SOLID)
            area.setPropertyValue("FillColor", int(0xEEEEEE))

    def test_wall(self):
        wall = self._diagram.getWall()
        # change background color of area
        wall.setPropertyValue("FillColor", int(0xCCCCCC))
        wall.setPropertyValue("FillStyle", SOLID)

    def test_title(self):
        self._chart_document.setPropertyValue("HasMainTitle", True)

        title = self._chart_document.getTitle()
        if title is not None:
            title.setPropertyValue("String", "Bar Chart in a Draw Document")

    def test_legend(self):
        legend = self._chart_document.getLegend()
        legend.setPropertyValue("Alignment", LEFT)
        legend.setPropertyValue("FillStyle", SOLID)
        legend.setPropertyValue("FillColor", int(0xEEDDEE))

    def test_threeD(self):
        self._diagram.setPropertyValue("Dim3D", True)
        self._diagram.setPropertyValue("Deep", True)
        # from Chart3DBarProperties:
        self._diagram.setPropertyValue("SolidType", int(CYLINDER))

        # change floor color to Magenta6
        diagram_floor = self._diagram.getFloor()
        diagram_floor.setPropertyValue("FillColor", int(0x6B2394))

        # apply changes to get a 3d scene
        self._chart_document.unlockControllers()
        self._chart_document.lockControllers()

        # rotate scene to a different angle
        matrix = HomogenMatrix()
        lines = [
            HomogenMatrixLine(1.0, 0.0, 0.0, 0.0),
            HomogenMatrixLine(0.0, 1.0, 0.0, 0.0),
            HomogenMatrixLine(0.0, 0.0, 1.0, 0.0),
            HomogenMatrixLine(0.0, 0.0, 0.0, 1.0),
        ]

        matrix.Line1 = lines[0]
        matrix.Line2 = lines[1]
        matrix.Line3 = lines[2]
        matrix.Line4 = lines[3]

        # rotate 10 degrees along the x axis
        angle = 10.0
        cos_x = math.cos(math.pi / 180.0 * angle)
        sin_x = math.sin(math.pi / 180.0 * angle)

        # rotate -20 degrees along the y axis
        angle = -20.0
        cos_y = math.cos(math.pi / 180.0 * angle)
        sin_y = math.sin(math.pi / 180.0 * angle)

        # rotate -5 degrees along the z axis
        angle = -5.0
        cos_z = math.cos(math.pi / 180.0 * angle)
        sin_z = math.sin(math.pi / 180.0 * angle)

        line1 = HomogenMatrixLine(cos_y * cos_z, cos_y * -sin_z, sin_y, 0.0)
        line2 = HomogenMatrixLine(
            sin_x * sin_y * cos_z + cos_x * sin_z,
            -sin_x * sin_y * sin_z + cos_x * cos_z,
            -sin_x * cos_y,
            0.0,
        )
        line3 = HomogenMatrixLine(
            -cos_x * sin_y * cos_z + sin_x * sin_z,
            cos_x * sin_y * sin_z + sin_x * cos_z,
            cos_x * cos_y,
            0.0,
        )
        matrix.Line1 = line1
        matrix.Line2 = line2
        matrix.Line3 = line3
        self._diagram.setPropertyValue("D3DTransformMatrix", matrix)

        # add a red light source

        # in a chart by default only the second (non-specular) light source is switched on
        # light source 1 is a specular light source
        self._diagram.setPropertyValue("D3DSceneLightColor1", int(0xFF3333))

        # set direction
        direction = Direction3D()
        direction.DirectionX = -0.75
        direction.DirectionY = 0.5
        direction.DirectionZ = 0.5

        self._diagram.setPropertyValue("D3DSceneLightDirection1", direction)
        self._diagram.setPropertyValue("D3DSceneLightOn1", True)


# Create a spreadsheet add some data and add a chart
def main():
    helper = Helper.Helper()
    chart_helper = ChartHelper.ChartHelper(helper.create_drawing_document())

    # the unit for measures is 1/100th of a millimeter
    # position at (1cm, 1cm)
    pos = Point(1000, 1000)

    # size of the chart is 15cm x 13cm
    extent = Size(15000, 13000)

    # insert a new chart into the "Chart" sheet of the
    # spreadsheet document
    chart_doc = chart_helper.insert_ole_chart_in_draw(
        pos, extent, "com.sun.star.chart.BarDiagram"
    )

    # instantiate test class with newly created chart
    test = ChartInDraw(chart_doc)

    try:
        test.lock_controllers()

        test.test_area()
        test.test_wall()
        test.test_title()
        test.test_legend()
        test.test_threeD()

        test.unlock_controllers()

    except Exception as err:
        print(f"UNO Exception caught: {err}")


# Main entry point
if __name__ == "__main__":
    main()

# vim: set shiftwidth=4 softtabstop=4 expandtab:
